tests/backends/vllm/test_migration.py

```diff
@@ -0,0 +1,137 @@
+                "Swahili: 'The early bird catches the worm.'\n"]
+
+class MockBackendVLLM(BackendVLLM):
+    def __init__(self):
+        self.engine = MockEngine()
+
+class MockLlumlet(Llumlet):
+    def __init__(self):
+        self.instance_id = "0"
+        self.backend_engine = MockBackendVLLM()
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2,
+                    reason="Need at least 2 GPUs to run the test.")
+def test_migration_correctness():
+    ray.init(namespace="llumnix", ignore_reinit_error=True)
+    engine_args = EngineArgs(model="facebook/opt-125m",worker_use_ray=True)
+    id_rank_map = {"0":0,"1":1}
+    migration_config = MigrationConfig("LCFS", "gloo",16,1,4,4,5,False)
+    que = RayQueue(actor_options={
+        "scheduling_strategy": NodeAffinitySchedulingStrategy(
+            node_id=ray.get_runtime_context().get_node_id(),
+            soft=False,)
+    })
+    server_info = ServerInfo("0",que)
+
+    llumlet_0:Llumlet = Llumlet.from_args(True,
+                      "0",
+                      BackendType.VLLM,
+                      1,
+                      migration_config,
+                      engine_args,
+                     )
+    llumlet_1:Llumlet = Llumlet.from_args(True,
+                      "1",
+                      BackendType.VLLM,
+                      1,
+                      migration_config,
+                      engine_args,
+                     )
+    while True:
+        res = ray.get([llumlet_0.is_ready.remote(),llumlet_1.is_ready.remote()])
+        if all(res):
+            break
+    ray.get([llumlet_0.execute_engine_method.remote("_run_workers","rebuild_migrate_backend", id_rank_map, "llumnix"),
+            llumlet_1.execute_engine_method.remote("_run_workers","rebuild_migrate_backend", id_rank_map, "llumnix")])
+    print("init done")
+    # empty instance migrate out
+    res = ray.get(llumlet_0.migrate_out.remote("instance_1"))
+    assert not res
+
+    # running without migration
+    def test_correctness(prompt):
+        sampling_params = SamplingParams(top_k=1, temperature=0, ignore_eos=True, max_tokens=100)
+        request_id0 = random_uuid()
+        llumlet_0.generate.remote(request_id0, server_info, prompt, sampling_params)
+        request_output_queue = que
+        origin_output = None
+        finished = False
+        while not finished:
+            qsize = ray.get(request_output_queue.actor.qsize.remote())
+            request_outputs = ray.get(request_output_queue.actor.get_nowait_batch.remote(qsize))
+            for request_output in request_outputs:
+                origin_output = request_output.outputs[0]
+                finished = request_output.finished
+
+        request_id1 = random_uuid()
+        llumlet_0.generate.remote(request_id1, server_info, prompt, sampling_params)
+        # wait prefill done
+        time.sleep(0.1)
```

Comment:
what if prefill does not finish after 0.1s? or what if the request has finished after 0.1s? seems there exists randomness in this test

Reply:
Here, using sleep does not seem to be a good method. We need a more explicit way to ensure that only part of the step is completed.

Reply:
Fixed by replacing `time.sleep` with `while True`